Add supervisor components alongside the current  agents

owtaylor opened the pull request:
Add a new workflow supervisor to handle coordinating the test and release of builds once the merge request made be the agents is merged.

There are two major components: the collector (queries JIRA and erratatool for what issues and errata need processing) and the executor (executes tasks).

 * README.md is split into README.md / README-agents.md / README.supervisor.md
 * compose.yaml now has two profiles "agents" and "supervisor"
 * A single image is used for both the agents and the supervisor

The structure here should be pretty workable, but the actual implementation of checks and actions is still undone to a large amount - in particular, *no* actions are implemented at all.

The integration into beeai/ feels a little intrusive - but I do think it makes sense to have a single codebase with both the existing agents and this:

 - They use common dependencies (beeai-framework, redis, etc.), and need much the same setup, utilities and tools.
 - We probably want to eventually orchestrate them in a single project to reduce the amount of "ops" overhead we have.
 - There may be opportunities in the future where we merge the two more - for example, we might want to use a Task as defined here to monitor tests on a gitlab MR we've created.

owtaylor commented on beeai/supervisor/main.py:
```diff
@@ -0,0 +1,189 @@
+def process_erratum(id_or_url: str):
+    if id_or_url.startswith("http"):
+        m = re.match(
+            r"https://errata.engineering.redhat.com/advisory/([\d+])$", id_or_url
```
Fixed

owtaylor commented on beeai/supervisor/main.py:
```diff
@@ -0,0 +1,189 @@
+                await collect_once(queue)
+            except Exception:
+                logger.exception("Error while collecting tasks")
+            await asyncio.sleep(20 * 60)
```
Fixed

owtaylor commented on beeai/Containerfile:
```diff
@@ -23,6 +26,10 @@ COPY beeai-gemini-malformed-function-call.patch /tmp
 COPY beeai-instrumentation.patch /tmp
 
+# Add Red Hat internal certificate authority
+ADD certs/Current-IT-Root-CAs.pem /etc/pki/ca-trust/source/anchors/
```
Fixed

halfline commented on beeai/supervisor/main.py:
```diff
@@ -119,18 +128,21 @@ def execute(repeat: bool = typer.Option(True)):
 
 
 async def do_process_issue(key: str):
+    assert app_state.dry_run
```
was this just here to prevent spamming prod? If so do you need to drop it?

halfline commented on beeai/supervisor/main.py:
```diff
@@ -119,18 +128,21 @@ def execute(repeat: bool = typer.Option(True)):
         key,
         result.status,
-        result.reschedule_in if result.reschedule_in > 0 else "never",
+        result.reschedule_in if result.reschedule_in >= 0 else "never",
```
unrelated bug fix that should probably be squashed into earlier commits

owtaylor replied:
Done

halfline commented on beeai/Containerfile:
```diff
@@ -20,6 +23,10 @@ COPY beeai-gemini-malformed-function-call.patch /tmp
 COPY beeai-propagate-mcp-errors.patch /tmp
 
+# Add Red Hat internal certificate authority
+COPY certs/Current-IT-Root-CAs.pem /etc/pki/ca-trust/source/anchors/
```
adding the internal cert to the container image isn't awesome but short of forwarding `CA_REQUESTS_BUNDLE` `SSL_CERT_FILE` etc i'm not sure of a better approach.

halfline commented on beeai/supervisor/main.py:
```diff
@@ -0,0 +1,189 @@
+    else:
+        logging.basicConfig(level=logging.INFO)
+
+    collector_endpoint = os.environ.get("COLLECTOR_ENDPOINT")
+    if collector_endpoint is not None:
+        setup_observability(collector_endpoint)
```
it's a little unfortunate we have two kinds of collectors here... i wonder if it would make sense to call the main on a "scanner" or "scheduler" or something to avoid the clash.

halfline commented on beeai/supervisor/jira_utils.py:
```diff
@@ -225,5 +233,99 @@ def get_issues_statuses(issue_keys: Collection[str]) -> dict[str, IssueStatus]:
+def add_issue_label(
+    issue_key: str, label: str, comment: CommentSpec = None, *, dry_run: bool = False
+) -> None:
+    url = f"https://issues.redhat.com/rest/api/2/issue/{issue_key}/labels"
```
is that url right? or should it be `    url = f"https://issues.redhat.com/rest/api/2/issue/{issue_key}"` ?

owtaylor commented on beeai/supervisor/jira_utils.py:
```diff
@@ -86,49 +97,98 @@ def custom_enum_list(enum_class: Type[_E], name) -> list[_E] | None:
-def get_issue(issue_key) -> Issue:
-    url = f"https://issues.redhat.com/rest/api/2/issue/{issue_key}?fields=*navigable"
+    if full:
+        return FullIssue(
+            key=key,
+            url=f"https://issues.redhat.com/browse/{key}",
+            summary=issue_data["fields"]["summary"],
+            status=issue_data["fields"]["status"]["name"],
+            components=issue_components,
+            fix_versions=[v["name"] for v in issue_data["fields"]["fixVersions"]],
+            errata_link=errata_link,
+            fixed_in_build=custom("Fixed in Build"),
+            test_coverage=custom_enum_list(TestCoverage, "Test Coverage"),
+            preliminary_testing=custom_enum(PreliminaryTesting, "Preliminary Testing"),
+            comments=[
+                IssueComment(
+                    authorName=c["author"]["displayName"],
+                    authorEmail=c["author"]["emailAddress"],
+                    created=datetime.fromisoformat(c["created"]),
+                    body=c["body"],
+                )
+                for c in issue_data["fields"]["comment"]["comments"]
+            ],
+        )
+    else:
+        return Issue(
+            key=key,
+            url=f"https://issues.redhat.com/browse/{key}",
+            summary=issue_data["fields"]["summary"],
+            status=issue_data["fields"]["status"]["name"],
+            components=issue_components,
+            fix_versions=[v["name"] for v in issue_data["fields"]["fixVersions"]],
+            errata_link=errata_link,
+            fixed_in_build=custom("Fixed in Build"),
+            test_coverage=custom_enum_list(TestCoverage, "Test Coverage"),
+            preliminary_testing=custom_enum(PreliminaryTesting, "Preliminary Testing"),
+        )
```
The problem with that is that you lose static type safety - type checkers will no longer know if issue_arguments correspond to what is needed for type safety. What you can do is:

```py
    issue = Issue(
        key=key,
        url=f"https://issues.redhat.com/browse/{urlquote(key)}",
        summary=issue_data["fields"]["summary"],
        status=issue_data["fields"]["status"]["name"],
        components=issue_components,
        fix_versions=[v["name"] for v in issue_data["fields"]["fixVersions"]],
        errata_link=errata_link,
        fixed_in_build=custom("Fixed in Build"),
        test_coverage=custom_enum_list(TestCoverage, "Test Coverage"),
        preliminary_testing=custom_enum(PreliminaryTesting, "Preliminary Testing"),
    )

    if full:
        return FullIssue(
            **issue.__dict__,
            comments=[
                IssueComment(
                    authorName=c["author"]["displayName"],
                    authorEmail=c["author"]["emailAddress"],
                    created=datetime.fromisoformat(c["created"]),
                    body=c["body"],
                )
                for c in issue_data["fields"]["comment"]["comments"]
            ],
        )
    else:
        return issue
```

Which is slightly less efficient and not perfectly type safe (it won't notice if new required fields are added to FullIssue), but probably is a good compromise.

halfline replied:
i see your point... what about consolidating Issue to be one class and just making the "fullness" be broken out as typestate classes...like ```class Abridged:
    """Issue is minimally loaded for fast access""

class Full:
    """Issue contains all comments and other important extended state""

then:

FullnessState = TypeVar("FullnessState", Abridged, Full)

class Issue(Generic[FullnessState]):
    def __init__(
        self,
        key: str,
        url: str,
        summary: str,
        status: str,
        components: list[str],
        fix_versions: list[str],
        errata_link: str | None,
        fixed_in_build: str | None,
        test_coverage: str | None,
        preliminary_testing: str | None,
        comments: Sequence[IssueComment] = ()
    ):
``` then make the type hints use  `Issue[Full]` or `Issue[Abridged]` instead of `Issue` and `FullIssue`

halfline commented on beeai/supervisor/main.py:
```diff
@@ -0,0 +1,189 @@
+@app.command()
+def collect(
+    repeat: bool = typer.Option(True, "--repeat"),
+    repeat_delay=typer.Option(1200, "--repeat-delay"),
```
```diff
-    repeat_delay=typer.Option(1200, "--repeat-delay"),
+    repeat_delay: int = typer.Option(1200, "--repeat-delay"),
```

owtaylor replied:
Fixed

owtaylor commented on beeai/supervisor/main.py:
```diff
@@ -70,7 +79,7 @@ async def execute_once(queue: TaskQueue):
     if task.task_type == TaskType.PROCESS_ISSUE:
         issue = get_issue(task.task_data, full=True)
-        result = await run_issue_workflow(issue)
+        result = await IssueWorkflow(issue, dry_run=app_state.dry_run).run()
```
Yeah, it's not great naming. In a local commit, I named the "Executor" component to "Processor", so I can't really call these processors too.

 Idea 1: IssueWorkflowStep
 Idea 2: IssueHandler
 Idea 2: IssueStepHandler

I think:
 
```
 await IssueHandler(issue, dry_run=app_state.dry_run).run_step()
```

reads pretty well.

owtaylor commented on beeai/supervisor/testing_analyst.py:
```diff
@@ -1,86 +1,122 @@
             execution=get_agent_execution_config(),
         )
-        return OutputSchema.model_validate_json(response.answer.text)
+        return OutputSchema.model_validate_json(response.result.text)
```
It was unclear, but I made the previous commit the one that  switched from RequirementAgent to ToolCallingAgent and mentioned it in the commit  message.

owtaylor commented on beeai/templates/supervisor.env:
```diff
@@ -0,0 +1,13 @@
+
+# API key or other credentials corresponding to the model above
+# https://console.cloud.google.com/apis/credentials -> Create credentials -> API key
+GEMINI_API_KEY=
```
Done.

owtaylor commented on beeai/supervisor/supervisor_types.py:
```diff
@@ -6,11 +6,12 @@
+    INTEGRATION = "Integration"  # RHEL only
+    RELEASE_PENDING = "Release Pending"  # RHEL only
+    DONE = "Done"  # RHEL ONLY
     CLOSED = "Closed"
```
I squashed this up to the initial commit and added a doc comment to Issue that describes RHEL vs. RHELMISC

owtaylor commented on beeai/compose.yaml:
```diff
@@ -77,18 +106,39 @@ services:
+
+  supervisor:
+    <<: *supervisor
+    command: ["sh", "-c", "echo 'This services isn't run directly"]
```
Fixed.

owtaylor commented on beeai/supervisor/jira_utils.py:
```diff
@@ -0,0 +1,165 @@
+@cache
+def jira_headers() -> dict[str, str]:
+    jira_token = os.environ["JIRA_TOKEN"]
+
```
OK, I borrowed envvar validation code from @majamassarini 's jira_issue_fetcher, so it will now fail:

```
ERROR:__main__:Missing required environment variables: JIRA_TOKEN
INFO:__main__:Required environment variables:
INFO:__main__:  CHAT_MODEL - name of model to use (e.g., gemini:gemini-2.5-pro))
INFO:__main__:  JIRA_URL - Jira instance URL (e.g., https://issues.redhat.com)
INFO:__main__:  JIRA_TOKEN - Jira authentication token
INFO:__main__:  REDIS_URL - Redis connection URL (e.g., redis://localhost:6379)
```

owtaylor commented on beeai/Makefile:
```diff
@@ -91,17 +92,48 @@ trigger-pipeline:
+supervisor-stop:
+	$(COMPOSE_SUPERVISOR) down
+
+.PHONY: supervisor-clean
+supervisor-clean:
 	$(COMPOSE) -f $(COMPOSE_FILE) down --volumes
```
Removed.

owtaylor commented on beeai/supervisor/jira_utils.py:
```diff
@@ -0,0 +1,163 @@
+            #                custom_fields["Test Coverage"],
+            #                custom_fields["Preliminary Testing"],
+            #                custom_fields["Due Date"],
+            #            ],
```
Testing with the 145 issues for
```
project = RHEL and "QA Contact" = "tpelka@redhat.com" and status not in (Done, Closed)
```

Fetching the Issues for them (results checked to be identical)

```
With selected fields: 0.7s 0.6s 0.6s
With all navigable fields: 15.9s 16.2s 16.3
```

So, maintaining the field list here seems to be worthwhile.

owtaylor replied:
With the verification that it's useful and correct, I made the field list used again instead of commented out.

owtaylor commented:
Pushed a new version:

 * Rebased to lastest main
 * Addressed all comments on PR
 * Renamed "Executor" => "Processor"
 * Renamed `IssueWorkflow`/`ErrataWorkflow` to `IssueHandler`/`ErrataHandler`
 * Fixed a couple of issues with command line parsing and URL quoting
 * Switched to mounting the Red Hat IT CA certificate rather than embedding it in the image
 * Rewrote the history to a clean ahistorical state without subsequent fixes or renames

I think this should be basically ready to merge at this point.

owtaylor commented:
Pushed:
 * Another rebase
 * Minor changes to work with the new image base of centos 10 / Python-3.12
 * Fix IT cert handling to work with gitlab.cee.redhat.com as well.

halfline commented on beeai/supervisor/jira_utils.py:
```diff
@@ -0,0 +1,172 @@
+@cache
+def get_custom_fields() -> dict[str, str]:
+    URL = "https://issues.redhat.com/rest/api/2/field"
+    response = requests.get(URL, headers=jira_headers())
```
another thought...should all these requests have timeouts in case jira is hung? I guess the service isn't going to be able to do much in that case so maybe doesn't matter.

owtaylor replied:
Yes, needs a raise_for_status(). (Fixed.) Beyond that, there's some stuff that could be done at the session level to add timeouts/retries/etc. We'll add that as we start having reliability issue :-)

halfline commented on beeai/supervisor/tools/read_readme.py:
```diff
@@ -0,0 +1,52 @@
+                                result=await response.text(),
+                            )
+        return StringToolOutput(
+            result=f"Failed to find RREADME.md for {input.repo_url}"
```
typo

halfline commented on beeai/supervisor/work_queue.py:
```diff
@@ -0,0 +1,111 @@
+    async def schedule_work_items(
+        self, work_items: Iterable[WorkItem], delay: float = 0.0
+    ) -> None:
+        new_time = time.time() + delay
```
of course if you did change, would have to change here too

owtaylor commented on beeai/supervisor/errata_utils.py:
```diff
@@ -0,0 +1,170 @@
+
+
+@cache
+def get_erratum(erratum_id: str | int):
```
Err, that's poor thinking, erratum_links is already deduplicated with a `set()`. I'll just remove all of the caching :-)

owtaylor commented on beeai/supervisor/erratum_handler.py:
```diff
@@ -0,0 +1,93 @@
+                dedent(
+                    f"""\
+                    Transition to {new_status} is blocked by:
+                    {"\n".join(f"{r.name}: {r.details}\n" for r in rule_set.rules if r.outcome == TransitionRuleOutcome.BLOCK)}
```
Yeah, fishy. Something like:
```
                    Transition to {new_status} is blocked by:\n
                    {"\n".join(f"{r.name}: {r.details}" for ....
```
is probably what I meant.
